"""
Standard tags.
"""

from enum import StrEnum, unique


@unique
class StandardTag(StrEnum):
    """
    The predefined standard tags.

    https://samtools.github.io/hts-specs/SAMtags.pdf
    """

    AM = "AM"
    """The smallest template-independent mapping quality in the template."""

    AS = "AS"
    """Alignment score generated by aligner."""

    BC = "BC"
    """Barcode sequence identifying the sample."""

    BQ = "BQ"
    """Offset to base alignment quality (BAQ)."""

    BZ = "BZ"
    """Phred quality of the unique molecular barcode bases in the OX tag."""

    CB = "CB"
    """Cell identifier."""

    CC = "CC"
    """Reference name of the next hit."""

    CG = "CG"
    """BAM only: CIGAR in BAMâ€™s binary encoding if (and only if) it consists of >65535 operators."""

    CM = "CM"
    """Edit distance between the color sequence and the color reference (see also NM)."""

    CO = "CO"
    """Free-text comments."""

    CP = "CP"
    """Leftmost coordinate of the next hit."""

    CQ = "CQ"
    """Color read base qualities."""

    CR = "CR"
    """Cellular barcode sequence bases (uncorrected)."""

    CS = "CS"
    """Color read sequence."""

    CT = "CT"
    """Complete read annotation tag, used for consensus annotation dummy features."""

    CY = "CY"
    """Phred quality of the cellular barcode sequence in the CR tag."""

    E2 = "E2"
    """The 2nd most likely base calls."""

    FI = "FI"
    """The index of segment in the template."""

    FS = "FS"
    """Segment suffix."""

    FZ = "FZ"
    """Flow signal intensities."""

    GC = "GC"
    """Reserved for backwards compatibility reasons."""

    GQ = "GQ"
    """Reserved for backwards compatibility reasons."""

    GS = "GS"
    """Reserved for backwards compatibility reasons."""

    H0 = "H0"
    """Number of perfect hits."""

    H1 = "H1"
    """Number of 1-difference hits (see also NM)."""

    H2 = "H2"
    """Number of 2-difference hits."""

    HI = "HI"
    """Query hit index."""

    IH = "IH"
    """Query hit total count."""

    LB = "LB"
    """Library."""

    MC = "MC"
    """CIGAR string for mate/next segment."""

    MD = "MD"
    """String encoding mismatched and deleted reference bases."""

    MF = "MF"
    """Reserved for backwards compatibility reasons."""

    MI = "MI"
    """Molecular identifier; a string that uniquely identifies the molecule from which the record was derived."""

    ML = "ML"
    """Base modification probabilities."""

    MM = "MM"
    """Base modifications / methylation."""

    MQ = "MQ"
    """Mapping quality of the mate/next segment."""

    NH = "NH"
    """Number of reported alignments that contain the query in the current record."""

    NM = "NM"
    """Edit distance to the reference."""

    OA = "OA"
    """Original alignment."""

    OC = "OC"
    """Original CIGAR (deprecated; use OA instead)."""

    OP = "OP"
    """Original mapping position (deprecated; use OA instead)."""

    OQ = "OQ"
    """Original base quality."""

    OX = "OX"
    """Original unique molecular barcode bases."""

    PG = "PG"
    """Program."""

    PQ = "PQ"
    """Phred likelihood of the template."""

    PT = "PT"
    """Read annotations for parts of the padded read sequence."""

    PU = "PU"
    """Platform unit."""

    Q2 = "Q2"
    """Phred quality of the mate/next segment sequence in the R2 tag."""

    QT = "QT"
    """Phred quality of the sample barcode sequence in the BC tag."""

    QX = "QX"
    """Quality score of the unique molecular identifier in the RX tag."""

    R2 = "R2"
    """Sequence of the mate/next segment in the template."""

    RG = "RG"
    """Read group."""

    RT = "RT"
    """Reserved for backwards compatibility reasons."""

    RX = "RX"
    """Sequence bases of the (possibly corrected) unique molecular identifier."""

    S2 = "S2"
    """Reserved for backwards compatibility reasons."""

    SA = "SA"
    """Other canonical alignments in a chimeric alignment."""

    SM = "SM"
    """Template-independent mapping quality."""

    SQ = "SQ"
    """Reserved for backwards compatibility reasons."""

    TC = "TC"
    """The number of segments in the template."""

    TS = "TS"
    """Transcript strand."""

    U2 = "U2"
    """Phred probability of the 2nd call being wrong conditional on the best being wrong."""

    UQ = "Phred likelihood of the segment, conditional on the mapping being correct."""
